pyquil/operator_estimation.py

```diff
@@ -0,0 +1,428 @@
+        return {
+            'type': 'ExperimentSuite',
+            'experiments': self._experiments,
+            'program': self.program.out(),
+            'qubits': self.qubits,
+        }
+
+    def __eq__(self, other):
+        if not isinstance(other, ExperimentSuite):
+            return False
+        return self.serializable() == other.serializable()
+
+
+class OperatorEncoder(JSONEncoder):
+    def default(self, o):
+        if isinstance(o, Experiment):
+            return str(o)
+        if isinstance(o, ExperimentSuite):
+            return {
+                'type': 'ExperimentSuite',
+                'experiments': o._experiments,
+                'program': o.program.out(),
+                'qubits': o.qubits,
+            }
+        if isinstance(o, ExperimentResult):
+            return {
+                'type': 'ExperimentResult',
+                'experiment': o.experiment,
+                'expectation': o.expectation,
+                'stddev': o.stddev,
+            }
+
+        return o
+
+
+def to_json(fn, obj):
+    with open(fn, 'w') as f:
+        json.dump(obj, f, cls=OperatorEncoder, indent=2)
+    return fn
+
+
+def _operator_object_hook(obj):
+    if 'type' in obj and obj['type'] == 'ExperimentSuite':
+        return ExperimentSuite([[Experiment.from_str(e) for e in exps]
+                                for exps in obj['experiments']],
+                               program=Program(obj['program']),
+                               qubits=obj['qubits'])
+    return obj
+
+
+def read_json(fn):
+    with open(fn) as f:
+        return json.load(f, object_hook=_operator_object_hook)
+
+
+def _local_pauli_eig_prep(op: str, idx: int):
+    """
+    Generate gate sequence to prepare a the +1 eigenstate of a Pauli operator, assuming
+    we are starting from the ground state (the +1 eigenstate of Z^{\\otimes n}).
+
+    :param op: A string representation of the Pauli operator whose eigenstate we'd like to prepare.
+    :param idx: The index of the qubit that the preparation is acting on
+    :return: A program which will prepare the requested state.
+    """
+    if op == 'X':
+        gate = RY(pi / 2, idx)
+    elif op == 'Y':
+        gate = RX(-pi / 2, idx)
+    elif op == 'Z':
+        return Program()
+    else:
+        raise ValueError('Unknown gate operation')
+
+    return Program(gate)
+
+
+def _local_pauli_eig_meas(op, idx):
+    """
+    Generate gate sequence to measure in the eigenbasis of a Pauli operator, assuming
+    we are only able to measure in the Z eigenbasis.
+
+    """
+    if op == 'X':
+        gate = RY(-pi / 2, idx)
+    elif op == 'Y':
+        gate = RX(pi / 2, idx)
+    elif op == 'Z':
+        return Program()
+    else:
+        raise ValueError('Unknown gate operation')
+    return Program(gate)
+
+
+def _ops_belong_to_a_tpb(op_code1: str, op_code2: str):
+    """
+    Given two op strings (I, X, Y, or Z) return whether they share a tensor product basis.
```

Comment:
Maybe we could add a qualifying statement here that says that operators always act on the full Hilbert space of all available qubits, to clear the confusion. But I agree that something like `_ops_commute(..)` is both more compact, and meaningful.

Reply:
I'm a little out of my depth here, but I need a term for "can use one set of 1q rotations to measure all these operators" which is *not* the same thing as whether observables commute. It's much more restricting. 

In grove, @ncrubin had a function which talked about whether operators "trivially commute" to mean the above, but I don't think this is standard terminology. I've seen TPB used for this, see for example section V.A. of https://arxiv.org/abs/1704.05018. 

@marcusps thoughts?

Reply:
You can always use *some* set of rotations (maybe not necessarily all of them 1q) to measure all operators. Two observables commute if and only if they are simultaneously diagonalizable, but our way of constructing the diagonalizing basis for some operator does not always guarantee that the other commuting observable will also be diagonal in that basis.

Maybe a better term for this could be `_ops_diagonal_in_a_tpb(..)`?

Reply:
The paper https://arxiv.org/pdf/1704.05018.pdf introduces this concept with the words "we group ... Pauli operators ... in ... sets ... which have terms that are *diagonal* in the same *tensor product basis*." So I still stick with the opinion that `_ops_diagonal_in_a_tpb(..)` is a better term to use here.

Reply:
changed to suggested name in 0b8a560ee469989a34029ac244846e8fd1790163